torch/csrc/jit/passes/utils/subgraph_matcher.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <torch/csrc/jit/ir.h>
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+
+// Usage: instantiate a Graph and use a MatchIterator to iteratively
+// get matches on the underlying graph.  The returned `Match`s can
+// be used as maps from the passed in match graph to the graph the
+// match was run on.
+// 
+//   for (auto m : MatchIterator(match_graph, graph)) {
+//     AT_ASSERT(m[m0.value()] == a.value());
+//   }
+//
+// Optionally, pass in a std::ostream* to get debug information.
+// e.g. `MatchIterator(match_graph, graph, &std::cout)`.
+//
+struct TORCH_API MatchIterator {
+  MatchIterator(
+      const Graph& match_graph,
+      const Graph& graph,
+      std::ostream* log_stream = nullptr)
+      : match_graph_(match_graph),
+        // Arbitrarily choose anchor node
+        anchor_(*(match_graph.nodes().begin())),
+        graph_(graph),
+        finished_(false),
+        log_stream_(log_stream) {}
+
+  // Match is a two typed map that makes it convenient to deal with
+  // the matches by overloading the subscript [] operator to be 
+  // const correct and handle both types (Node*, Value*)
+  struct Match {
+    std::unordered_map<const Node*, const Node*> node_match;
+    std::unordered_map<const Value*, const Value*> value_match;
+
+    size_t size() const;
+    void clear();
+
+    void set(const Node* const& m, const Node* n);
+    void set(const Value* m, const Value* v);
+
+    const Node* get(const Node* const& m) const;
+    const Value* get(const Value* const& m) const;
+    const Node* operator[](const Node* const& m) const;
+    const Value* operator[](const Value* const& m) const;
+  };
+
+  MatchIterator& begin();
+  MatchIterator& end();
+
+  MatchIterator& operator++();
+  MatchIterator& operator++(int);
+
+  Match operator*();
+  bool operator!=(MatchIterator& rhs);
+
+ private:
+  using NodeFrontier = std::vector<std::pair<const Node*, const Node*>>;
+  using ValueFrontier = std::vector<std::pair<const Value*, const Value*>>;
+
+  bool matchNodes(const Node* n1, const Node* n2) {
+    auto match = n1->kind() == n2->kind();
+    std::stringstream os;
+    if (!match) {
+      os << "[mismatch] node types don't match: "
+         << n1->kind().toDisplayString() << "(" << n1 << ")"
+         << " vs "
+         << n2->kind().toDisplayString() << "(" << n2 << ")";
+    } else {
+      os << "[match] node of type " << n1->kind().toDisplayString()
+         << " " << n1 << " - " << n2;
+    }
+    log(os.str());
+    return match;
+  }
+  bool matchValues(const Value* v1, const Value* v2) {
```

Comment:
Why is there a function that returns true?

Reply:
mostly to capture the logging event (which node was mapped to which)